zeppelin-zengine/src/main/java/org/apache/zeppelin/scheduler/pool/DynamicThreadPool.java

```diff
@@ -0,0 +1,248 @@
+ *
+ * Instance could be obtained using <code>{@link DynamicThreadPool#getInstance(String)}</code>
+ */
+public class DynamicThreadPool implements ThreadPool {
+  private static final Logger LOGGER = LoggerFactory.getLogger(DynamicThreadPool.class);
+
+  private static final Map<String, DynamicThreadPool> INSTANCES = new ConcurrentHashMap<>();
+  private final Object poolEvent = new Object();
+
+  private final ThreadPoolExecutor executor =
+      new ThreadPoolExecutor(1, 10, 1, TimeUnit.MINUTES, new SynchronousQueue<>());
+
+  private volatile int priority = Thread.NORM_PRIORITY;
+
+  private String threadNamePrefix;
+  private String schedulerInstanceId;
+  private String schedulerInstanceName;
+
+  public static DynamicThreadPool getInstance(String schedulerInstanceId) {
+    return INSTANCES.get(schedulerInstanceId);
+  }
+
+  @Override
+  public boolean runInThread(Runnable runnable) {
+    if (runnable == null) {
+      return false;
+    }
+    try {
+      if (blockForAvailableThreads() > 0) {
+        executor.execute(new NotifableJob(runnable));
+      }
+      return true;
+    } catch (RejectedExecutionException e) {
+      LOGGER.error(e.toString());
+      return false;
+    }
+  }
+
+  @Override
+  public int blockForAvailableThreads() {
+    int availableThreads = executor.getMaximumPoolSize() - executor.getActiveCount();
+    while (availableThreads <= 0) {
+      synchronized (poolEvent) {
+        try {
+          poolEvent.wait(500);
+        } catch (InterruptedException e) {
+          Thread.currentThread().interrupt();
+          return 0;
+        }
+      }
+      availableThreads = executor.getMaximumPoolSize() - executor.getActiveCount();
+    }
+    return availableThreads;
+  }
+
+
+  @Override
+  public void initialize() throws SchedulerConfigException {
+    if (schedulerInstanceId != null && INSTANCES.get(schedulerInstanceId) != this) {
+      // already initialized...
+      return;
+    }
+
+    if (priority <= 0 || priority > 9) {
+      throw new SchedulerConfigException(
+          "Thread priority must be > 0 and <= 9");
+    }
+
+    if (threadNamePrefix == null && schedulerInstanceName != null) {
+      threadNamePrefix = schedulerInstanceName;
+    }
+    executor.setThreadFactory(new DynamicThreadFactory());
+  }
+
+  /**
+   * Terminate any worker threads in this thread group.
+   *
+   * Jobs currently in progress will complete.
+   *
+   * @param waitForJobsToComplete boolean flag for resource release: if true: initiates an orderly
+   * shutdown in which previously submitted tasks are executed, but no new tasks will be accepted.
+   * @see ExecutorService#shutdown() if false: Attempts to stop all actively executing tasks, halts
+   * the processing of waiting tasks.
+   * @see ExecutorService#shutdownNow()
+   */
+  @Override
+  public void shutdown(boolean waitForJobsToComplete) {
+    LOGGER.info("Shutting down ThreadPool...");
+    if (!waitForJobsToComplete) {
```

Comment:
```suggestion
    if (waitForJobsToComplete) {
```
Перепутана логика `waitForJobsToComplete`.

Reply:
Упс, это я так `executor.shutdownNow()` тестил и забыл вернуть :man_facepalming: